src/prof_recent.c

```diff
@@ -0,0 +1,328 @@
+	malloc_mutex_lock(tsd_tsdn(tsd), tctx->tdata->lock);
+	assert(tctx->recent_count > 0);
+	--tctx->recent_count;
+	if (prof_tctx_should_destroy(tsd_tsdn(tsd), tctx)) {
+		prof_tctx_destroy(tsd, tctx);
+	} else {
+		malloc_mutex_unlock(tsd_tsdn(tsd), tctx->tdata->lock);
+	}
+}
+
+void
+prof_recent_alloc_reset(tsd_t *tsd, const extent_t *extent) {
+	malloc_mutex_lock(tsd_tsdn(tsd), &prof_recent_alloc_mtx);
+	prof_recent_t *recent = extent_prof_recent_alloc_get(extent);
+	malloc_mutex_unlock(tsd_tsdn(tsd), &prof_recent_alloc_mtx);
+
+	if (recent == NULL) {
+		return;
+	}
+
+	prof_tctx_t *dalloc_tctx = prof_tctx_create(tsd);
+	if (dalloc_tctx != NULL) {
+		++dalloc_tctx->recent_count;
+		dalloc_tctx->prepared = false;
+	}
+
+	malloc_mutex_lock(tsd_tsdn(tsd), &prof_recent_alloc_mtx);
+	recent = extent_prof_recent_alloc_get(extent);
+	if (recent != NULL) {
+		recent->alloc_extent = NULL;
+		recent->dalloc_tctx = dalloc_tctx;
+		nstime_t dalloc_time = NSTIME_ZERO_INITIALIZER;
+		nstime_update(&dalloc_time);
+		recent->dalloc_time = nstime_ns(&dalloc_time);
+	} else if (dalloc_tctx != NULL) {
+		/* We lost the rase - the allocation record was just gone. */
+		decrement_recent_count(tsd, dalloc_tctx);
+	}
+	malloc_mutex_unlock(tsd_tsdn(tsd), &prof_recent_alloc_mtx);
+}
+
+static void
+notify_alloc_extent(tsd_t *tsd, prof_recent_t *recent) {
+	malloc_mutex_assert_owner(tsd_tsdn(tsd), &prof_recent_alloc_mtx);
+	if (recent->alloc_extent != NULL) {
+		extent_prof_recent_alloc_set(recent->alloc_extent, NULL);
+	}
+}
+
+void
+prof_recent_alloc(tsd_t *tsd, prof_tctx_t *tctx, const void *ptr, size_t usize,
+    nstime_t t) {
+	extent_t *extent = iealloc(tsd_tsdn(tsd), ptr);
+	assert(extent != NULL);
+	assert(tctx == extent_prof_tctx_get(extent));
+
+	malloc_mutex_assert_not_owner(tsd_tsdn(tsd), tctx->tdata->lock);
+	malloc_mutex_lock(tsd_tsdn(tsd), &prof_recent_alloc_mtx);
+
+	if (prof_recent_alloc_max == 0) {
+		goto label_rollback;
+	}
+
+	prof_recent_t *node;
+	prof_tctx_t *old_alloc_tctx, *old_dalloc_tctx;
+
+	if (prof_recent_alloc_max != -1 &&
+	    prof_recent_alloc_count == prof_recent_alloc_max) {
+		/* If the upper limit is reached, simply reuse the head node. */
+		assert(prof_recent_alloc_head != NULL);
+		assert(prof_recent_alloc_tail != NULL);
+		assert(prof_recent_alloc_count > 0);
+		node = prof_recent_alloc_head;
+		prof_recent_alloc_head = node->next;
+		if (prof_recent_alloc_head == NULL) {
+			prof_recent_alloc_tail = NULL;
+		}
+		--prof_recent_alloc_count;
+		old_alloc_tctx = node->alloc_tctx;
+		old_dalloc_tctx = node->dalloc_tctx;
+		notify_alloc_extent(tsd, node);
+	} else {
+		/* Otherwise allocate a new node. */
+		assert(prof_recent_alloc_max == -1 ||
+		    prof_recent_alloc_count < prof_recent_alloc_max);
+		node = (prof_recent_t *)iallocztm(tsd_tsdn(tsd),
+		    sizeof(prof_recent_t), sz_size2index(sizeof(prof_recent_t)),
+		    false, NULL, true, arena_get(TSDN_NULL, 0, true), true);
```

Comment:
`arena_get(tsd, 0, false)` is fine.

Reply:
I think you meant `arena_get(tsd_tsdn(tsd), 0, false)`? Also changed the `arena_get()` call for allocating the buffer for dumping.